src/serializers/fields.rs

```diff
@@ -242,24 +361,67 @@ impl GeneralFieldsSerializer {
         // we don't both with `used_fields` here because on unions, `to_python(..., mode='json')` is used
         let mut map = serializer.serialize_map(Some(expected_len))?;
 
-        for result in main_iter {
-            let (key, value) = result.map_err(py_err_se_err)?;
-            if extra.exclude_none && value.is_none() {
-                continue;
+        if !extra.sort_keys {
+            for result in main_iter {
+                let (key, value) = result.map_err(py_err_se_err)?;
+                let key_str = key_str(&key).map_err(py_err_se_err)?;
+                self.process_field_entry::<S>(key_str, &key, &value, &mut map, include, exclude, &extra)?;
             }
-            let key_str = key_str(&key).map_err(py_err_se_err)?;
-            let field_extra = Extra {
-                field_name: Some(key_str),
-                ..extra
-            };
+        } else {
+            let mut items = main_iter
+                .map(|r| -> PyResult<_> {
+                    let (k, v) = r?;
+                    let k_str = key_str(&k)?.to_owned();
+                    Ok((k_str, k, v))
+                })
+                .collect::<PyResult<Vec<_>>>()
+                .map_err(py_err_se_err)?;
+            items.sort_by(|(a, _, _), (b, _, _)| a.cmp(b));
+            for (key_str, key, value) in items {
+                self.process_field_entry::<S>(&key_str, &key, &value, &mut map, include, exclude, &extra)?;
+            }
+        }
+        Ok(map)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn process_field_entry<'py, S: serde::ser::Serializer>(
+        &self,
+        key_str: &str,
+        key: &Bound<'py, PyAny>,
+        value: &Bound<'py, PyAny>,
+        map: &mut S::SerializeMap,
+        include: Option<&Bound<'py, PyAny>>,
+        exclude: Option<&Bound<'py, PyAny>>,
+        extra: &Extra,
+    ) -> Result<(), S::Error> {
+        if extra.exclude_none && value.is_none() {
+            return Ok(());
+        }
+        let field_extra = Extra {
+            field_name: Some(key_str),
+            ..*extra
+        };
 
-            let filter = self.filter.key_filter(&key, include, exclude).map_err(py_err_se_err)?;
-            if let Some((next_include, next_exclude)) = filter {
-                if let Some(field) = self.fields.get(key_str) {
-                    if let Some(ref serializer) = field.serializer {
-                        if !exclude_default(&value, &field_extra, serializer).map_err(py_err_se_err)? {
+        let filter = self.filter.key_filter(key, include, exclude).map_err(py_err_se_err)?;
+        if let Some((next_include, next_exclude)) = filter {
+            if let Some(field) = self.fields.get(key_str) {
+                if let Some(ref serializer) = field.serializer {
+                    if !exclude_default(value, &field_extra, serializer).map_err(py_err_se_err)? {
+                        if extra.sort_keys && Self::contains_dict(value) {
```

Comment:
@adriangb , we use `contains_dict` check here as well. This is to ensure to check for any dictionary before we commit to actually try to sort.